distributed/tests/test_scheduler.py

```diff
@@ -3401,6 +3402,91 @@ async def test_TaskState__to_dict(c, s):
     assert tasks["y"]["dependencies"] == ["<TaskState 'x' no-worker>"]
 
 
+def _verify_cluster_state(
+    state: dict, workers: Collection[Worker], allow_missing: bool = False
+) -> None:
+    addrs = {w.address for w in workers}
+    assert state.keys() == {"scheduler", "workers", "versions"}
+    assert state["workers"].keys() == addrs
+    if allow_missing:
+        assert state["versions"]["workers"].keys() <= addrs
+    else:
+        assert state["versions"]["workers"].keys() == addrs
+
+
+@gen_cluster(nthreads=[("", 1)] * 2)
+async def test_get_cluster_state(s: Scheduler, *workers: Worker):
+    state = await s.get_cluster_state([])
+    _verify_cluster_state(state, workers)
+
+    await asyncio.gather(*(w.close() for w in workers))
+
+    while s.workers:
+        await asyncio.sleep(0.01)
+
+    state_no_workers = await s.get_cluster_state([])
+    _verify_cluster_state(state_no_workers, [])
+
+
+@gen_cluster(
+    nthreads=[("", 1)] * 2,
+    config={"distributed.comm.timeouts.connect": "200ms"},
+)
+async def test_get_cluster_state_worker_error(s: Scheduler, a: Worker, b: Worker):
+    a.stop()
+    state = await s.get_cluster_state([])
+    _verify_cluster_state(state, [a, b], allow_missing=True)
+    assert state["workers"][a.address] == (
+        f"OSError('Timed out trying to connect to {a.address} after 0.2 s')"
+    )
+    assert isinstance(state["workers"][b.address], dict)
+    assert state["versions"]["workers"].keys() == {b.address}
+
+
+def _verify_cluster_dump(url, format: str, workers: Collection[Worker]) -> dict:
+    import fsspec
+
+    if format == "msgpack":
+        import msgpack
+
+        url += ".msgpack.gz"
+        loader = msgpack.unpack
+
+    else:
+        import yaml
+
+        url += ".yaml"
+        loader = yaml.safe_load
+
+    with fsspec.open(url, mode="rb", compression="infer") as f:
+        state = loader(f)
+
+    _verify_cluster_state(state, workers)
+    return state
+
+
+@pytest.mark.parametrize("format", ["msgpack", "yaml"])
+@gen_cluster(nthreads=[("", 1)] * 2)
+async def test_dump_cluster_state(s: Scheduler, *workers: Worker, format):
+    fsspec = pytest.importorskip("fsspec")
+    try:
+        await s.dump_cluster_state_to_url(
+            "memory://state-dumps/two-workers", [], format
+        )
+        _verify_cluster_dump("memory://state-dumps/two-workers", format, workers)
+
+        await asyncio.gather(*(w.close() for w in workers))
+
+        while s.workers:
+            await asyncio.sleep(0.01)
+
+        await s.dump_cluster_state_to_url("memory://state-dumps/no-workers", [], format)
+        _verify_cluster_dump("memory://state-dumps/no-workers", format, [])
+    finally:
+        fs = fsspec.filesystem("memory")
+        fs.rm("state-dumps", recursive=True)
+
```

Comment:
For this test, I'd really like to confirm:
- [ ] that storage options are actually passed into fsspec
- [ ] that the writer doesn't block the event loop

But I couldn't think of a good way to do that, beyond mocking some fsspec internals.

What would probably work best is creating a dummy, pass-through fsspec filesystem implementation (where we can add an assertion about the `storage_options`, and a `time.sleep(1)` during the `read()`), then adding it to the URL with [url chaining](https://filesystem-spec.readthedocs.io/en/latest/features.html#url-chaining). But I don't know fsspec well, and I found the docs and implementation too confusing to be able to rig something up quickly.

If someone like @jrbourbeau could implement something like this quickly, that could be nice, otherwise I think we should just not worry about testing these things and spend our effort elsewhere.

Reply:
Peaking over the implementation, implementing a full filesystem is likely out of scope but something like this should work

```python
from fsspec.implementations.memory import MemoryFileSystem
from fsspec.registry import register_implementation

class SlowMemoryFileSystem(MemoryFileSystem):
    def open(self, *args, **kwargs):
        import time
        time.sleep(1)
        return super().open(*args, **kwargs)
register_implementation("slowmemory", SlowMemoryFileSystem)
```

then you can use `slowmemory` instead of `memory`. Modifying the actual write is not as easy. 

This test would require you to move the `open` into the thread as well but I don't see a problem with this.

Downside: The fsspec registry is readonly. I don't see any method to unregister this again